haystack/nodes/prompt/invocation_context_mapper.py

```diff
@@ -0,0 +1,313 @@
+        :param params: maps the function's input kwargs to some fixed values. For example `expand_value_to_list` expects
+            `value` and `target_list` parameters, so `params` might contain
+            `{'value': 'A', 'target_list': [1, 1, 1, 1]}` and the node will output `["A", "A", "A", "A"]`.
+            It does not need to contain all keyword args, see `inputs`.
+            You can use params to provide fallback values for arguments of `run` that you're not sure they will exist.
+            So if you need `query` to exist, you can provide a fallback value in the params, which will be used only if `query`
+            is not passed to this node by the pipeline.
+        :param outputs: under which key to store the outputs in the invocation context. The lenght of the outputs must match
+            the number of outputs produced by the function invoked.
+        """
+        super().__init__()
+        self.function = REGISTERED_FUNCTIONS[func]
+        self.outputs = outputs
+        self.inputs = inputs or {}
+        self.params = params or {}
+
+    def run(  # type: ignore
+        self,
+        query: Optional[str] = None,
+        file_paths: Optional[List[str]] = None,
+        labels: Optional[MultiLabel] = None,
+        documents: Optional[List[Document]] = None,
+        meta: Optional[dict] = None,
+        invocation_context: Optional[Dict[str, Any]] = None,
+    ) -> Tuple[Dict, str]:
+
+        # The invocation context overwrites locals(), so if for example invocation_context contains a
+        # modified list of Documents under the `documents` key, such list is used.
+        invocation_context = {**locals(), **(invocation_context or {})}
+        invocation_context.pop("self")
+        invocation_context = {key: value for key, value in invocation_context.items() if value is not None}
+
+        input_values = {
+            key: invocation_context[value]
+            for key, value in self.inputs.items()
+            if value in invocation_context.keys() and value is not None
+        }
+
+        input_values = {**self.params, **input_values}
+        try:
+            logger.debug(
+                "InvocationContextMapper is invoking this function: %s(%s)",
+                self.function.__name__,
+                ", ".join([f"{key}={value}" for key, value in input_values.items()]),
+            )
+            output_values = self.function(**input_values)
+        except TypeError as e:
+            raise ValueError(
+                "InvocationContextMapper could not apply the function to your inputs and parameters. "
+                "Check the above stacktrace and make sure you provided all the correct inputs, parameters, "
+                "and parameter types."
+            ) from e
+
+        for output_key, output_value in zip(self.outputs, output_values):
+            invocation_context[output_key] = output_value
+
+        output: Dict[str, Any] = {}
+        if query:
```

Comment:
Do we really need these assignment statements (281-290)? Seem like something that should be done by the pipeline module invoking run.

Reply:
Right, I'll double check and remove them :+1: